luma/core/device.py

```diff
@@ -250,13 +254,31 @@ def display(self, image):
 
         :param image: Image to display.
         :type image: PIL.Image.Image
+        :param framebuffer: Framebuffering strategy, currently instances of
+            ``diff_to_previous`` or ``full_frame`` are only supported.
         """
         assert image.mode == self.mode
         assert image.size == self.size
 
         image = self.preprocess(image)
         path = f"/dev/fb{self.id}"
-        data = bytes(self.__image_converter(image))
 
-        with open(path, "wb") as fp:
-            fp.write(data)
+        bytes_per_pixel = self.bits_per_pixel // 8
+        image_bytes_per_row = self.width * bytes_per_pixel
+
+        with open(path, "w+b") as fp:
+            with mmap.mmap(fp.fileno(), self.width * self.height * bytes_per_pixel,
+                           flags=mmap.MAP_SHARED, access=mmap.PROT_WRITE) as mm:
```

Comment:
Could the fp and memory-mapping be performed once?  I don't know if the expected sequence of calls make for any other good opportunity or if display() could just perform that work once, holding on to the file pointer and mapping for all subsequent calls?

Not that more speed is necessarily called for.

Reply:
I did try that initially, and noticed that the frame rate was initially lower (25 FPS vs. 30 FPS) and observed it degenerate over time (tending towards 20 FPS) over several runs of the bounce.py demo. I don’t know why this would be, as one might expect it to be marginally quicker. I did wonder if it was down to using instance variables rather than local variables (unlikely) but I have a feeling it down to mmap performance somehow (https://stackoverflow.com/a/38682857) or my overuse of yield/generators. I might try and rework it again to just use file `seek`ing instead to eliminate whether mmap is the culprit.

Another side effect of initialising fp and mm once is that there would be more tests that would need to be repaired, but I guess that is just me being lazy...

Reply:
Ok, so seeking and writing to the file pointer directly (rather than going through mmap), and the frame rate jump to 68 FPS, so I think we can saffely abandon mmap altogether

Reply:
Keeping the file handle open: frame rate goes to 73 FPS